apis/config/v1alpha2/configuration_types.go

```diff
@@ -27,6 +27,10 @@ import (
 type Configuration struct {
 	metav1.TypeMeta `json:",inline"`
 
+	// Namespace is the namespace in which kueue is deployed and is also used as part of DNSName.
+	// Defaults to kueue-system.
+	Namespace *string `json:"namespace,omitempty"`
```

Comment:
Forget to comment here, at the very beginning I was cheated by the annotation, actually we should name this serviceNamespace. What's more, I think just modify the api is not enough, I'm making some experiments  right now, I think we should also update the kustomize configuration.

Reply:
Maybe you can try with a different service rather than `webhook-service`, and apply an unqualified object. Then you will get an error like `failed to call webhook: Post "https://webhook-service.kueue-system.svc:443/mutate-kueue-x-k8s-io-v1alpha2-resourceflavor?timeout=10s": service "webhook-service" not found`.

Reply:
Because we still need to update the manifests.yaml. IIUC.

Reply:
What manifest did you test with? @kerthcet 
On my local test, it works fine.

- `tmp/tmp.yaml`

```yaml
kueue.x-k8s.io/v1alpha2
kind: ResourceFlavor
metadata:
  name: default
---
apiVersion: kueue.x-k8s.io/v1alpha2
kind: ClusterQueue
metadata:
  name: cluster-total
spec:
  namespaceSelector: {}
  resources:
  - name: "cpu"
    flavors:
    - name: default
      quota:
# invalid
        invalid: 9
  - name: "memory"
    flavors:
    - name: default
      quota:
        min: 36Gi
---
apiVersion: kueue.x-k8s.io/v1alpha2
kind: LocalQueue
metadata:
  namespace: default
  name: main
spec:
  clusterQueue: cluster-total
```

```shell
$ kubectl apply -k github.com/tenzen-y/kueue/config/default?ref=add-cert-options-to-component-config-customize-cert-opts
namespace/kueue-tenant-a created
customresourcedefinition.apiextensions.k8s.io/clusterqueues.kueue.x-k8s.io created
customresourcedefinition.apiextensions.k8s.io/localqueues.kueue.x-k8s.io created
customresourcedefinition.apiextensions.k8s.io/resourceflavors.kueue.x-k8s.io created
customresourcedefinition.apiextensions.k8s.io/workloads.kueue.x-k8s.io created
serviceaccount/kueue-controller-manager created
role.rbac.authorization.k8s.io/kueue-leader-election-role created
clusterrole.rbac.authorization.k8s.io/kueue-batch-admin-role created
clusterrole.rbac.authorization.k8s.io/kueue-batch-user-role created
clusterrole.rbac.authorization.k8s.io/kueue-clusterqueue-editor-role created
clusterrole.rbac.authorization.k8s.io/kueue-clusterqueue-viewer-role created
clusterrole.rbac.authorization.k8s.io/kueue-job-editor-role created
clusterrole.rbac.authorization.k8s.io/kueue-job-viewer-role created
clusterrole.rbac.authorization.k8s.io/kueue-localqueue-editor-role created
clusterrole.rbac.authorization.k8s.io/kueue-localqueue-viewer-role created
clusterrole.rbac.authorization.k8s.io/kueue-manager-role created
clusterrole.rbac.authorization.k8s.io/kueue-metrics-reader created
clusterrole.rbac.authorization.k8s.io/kueue-proxy-role created
clusterrole.rbac.authorization.k8s.io/kueue-resourceflavor-editor-role created
clusterrole.rbac.authorization.k8s.io/kueue-resourceflavor-viewer-role created
clusterrole.rbac.authorization.k8s.io/kueue-workload-editor-role created
clusterrole.rbac.authorization.k8s.io/kueue-workload-viewer-role created
rolebinding.rbac.authorization.k8s.io/kueue-leader-election-rolebinding created
clusterrolebinding.rbac.authorization.k8s.io/kueue-manager-rolebinding created
clusterrolebinding.rbac.authorization.k8s.io/kueue-proxy-rolebinding created
configmap/kueue-manager-config created
secret/kueue-tenant-a-secret created
service/kueue-controller-manager-metrics-service created
service/kueue-tenant-a-service created
deployment.apps/kueue-controller-manager created
mutatingwebhookconfiguration.admissionregistration.k8s.io/kueue-mutating-webhook-configuration created
validatingwebhookconfiguration.admissionregistration.k8s.io/kueue-validating-webhook-configuration created
$ k get pods -n kueue-tenant-a 
NAME                                        READY   STATUS    RESTARTS   AGE
kueue-controller-manager-698469f7c5-tx4nm   2/2     Running   0          58s
$ k get service -n kueue-tenant-a 
NAME                                       TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)    AGE
kueue-controller-manager-metrics-service   ClusterIP   10.96.131.29   <none>        8443/TCP   2m2s
kueue-tenant-a-service                     ClusterIP   10.96.122.75   <none>        443/TCP    2m2s
$ k apply -f tmp/tmp.yaml 
resourceflavor.kueue.x-k8s.io/default created
error: error validating "tmp/tmp.yaml": error validating data: ValidationError(ClusterQueue.spec.resources[0].flavors[0].quota): unknown field "invalid" in io.x-k8s.kueue.v1alpha2.ClusterQueue.spec.resources.flavors.quota; if you choose to ignore these errors, turn validation off with --validate=false
```

Reply:
Some conclusions:
1. When set the namespace, we should also update `config/default/kustomization.yaml`

2. When set serviceName, we should update `config/components/webhook/service.yaml` and `config/components/webhook/manifests.yaml` manually. I didn't find how to auto-generate the serviceName actually. I think there should be ways.

3. When set secretName, we should update `config/components/internalcert/secret.yaml` and `config/default/manager_webhook_patch.yaml`

Reply:
> I would put it inside the internalCertManagement struct

Actually the `namespace` helps to control where to deploy the kueue components, besides configure this field, we should also update the `config/default/kustomization.yaml`, to change the value of `namespace`. 

So I think we can leave it outside the internalCertManagement

Reply:
> Some conclusions:
>
> 1. When set the namespace, we should also update config/default/kustomization.yaml
>
> 2. When set serviceName, we should update config/components/webhook/service.yaml and config/components/webhook/manifests.yaml manually. I didn't find how to auto-generate the serviceName actually. I think there should be ways.
>
> 3. When set secretName, we should update config/components/internalcert/secret.yaml and > config/default/manager_webhook_patch.yaml
>
> Better to document them or we can find ways to automate it.

Thanks for clarifying! @kerthcet 
After some research, it seems that we just need to update `config/default/kustomization.yaml` with the kustomize [`replacements`](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/replacements/) feature in cases 2 and 3 as I show [this commit](https://github.com/kubernetes-sigs/kueue/pull/362/commits/20872ca4e8e7c09247e13ec5014a82e880c459f9). This method is convenient for users because they only need to update one file.

```diff
+replacements:
+  - source:
+      version: v1
+      kind: Secret
+      # If you change `metadata.name` in `config/components/secret.yaml` and
+      # `internalCertManagement.secretName` in `config/components/manager/controller_manager_config.yaml`,
+      # you should update the below name.
+      name: webhook-server-cert
+      fieldPath: metadata.name
+    targets:
+      - select:
+          group: apps
+          kind: Deployment
+          name: controller-manager
+        fieldPaths:
+          - spec.template.spec.volumes.0.secret.secretName
+        options:
+          create: true
+  - source:
+      version: v1
+      kind: Service
+      # If you change `metadata.name` in `config/components/secret.yaml` and
+      # `internalCertManagement.serviceName` in `config/components/manager/controller_manager_config.yaml`,
+      # you should update the below name.
+      name: webhook-service
+      fieldPath: metadata.name
+    targets:
+      - select:
+          group: admissionregistration.k8s.io
+          kind: MutatingWebhookConfiguration
+        fieldPaths:
+          - webhooks.0.clientConfig.service.name
+          - webhooks.1.clientConfig.service.name
+          - webhooks.2.clientConfig.service.name
+        options:
+          create: true
+      - select:
+          group: admissionregistration.k8s.io
+          kind: ValidatingWebhookConfiguration
+        fieldPaths:
+          - webhooks.0.clientConfig.service.name
+          - webhooks.1.clientConfig.service.name
+          - webhooks.2.clientConfig.service.name
+          - webhooks.3.clientConfig.service.name
+        options:
+          create: true

 # Adds namespace to all resources.
 namespace: kueue-system
...
```

In case 1, we can not overwrite structured data in ConfigMap using the kustomize configurations yet, but KEP has been submitted in [this PR](https://github.com/kubernetes-sigs/kustomize/pull/4558) and is under discussion.

So, as you say, it might be a good idea to mention in the documentation to update `config/default/kustomization.yaml` when setting namespace, secretName, and serviceName.

What do you think about using kustomize `replacements` feature? @kerthcet

Reply:
> Actually the namespace helps to control where to deploy the kueue components, besides configure this field, we should also update the config/default/kustomization.yaml, to change the value of namespace.
>
> So I think we can leave it outside the internalCertManagement

Thanks for your suggestions. I agree with you.

Reply:
I'm not quite familiar with Kustomize grammars, so I'd like to leave it to @alculquicondor , in case giving some bad advices. But I'm always for automation.

Reply:
This conversation has derailed in ways I don't understand :)

I don't think it's a primary use case to change the namespace name or service name, so I think we can leave it out of the documentation. IMO, it suffices to leave the hooks in the configuration to let users change the parameters without having to rebuild the image, **if they really need to**.

Reply:
> I don't think it's a primary use case to change the namespace name or service name, so I think we can leave it out of the documentation. IMO, it suffices to leave the hooks in the configuration to let users change the parameters without having to rebuild the image, if they really need to.

@alculquicondor That makes sense.
I will move [these configurations](https://github.com/kubernetes-sigs/kueue/blob/20872ca4e8e7c09247e13ec5014a82e880c459f9/config/default/kustomization.yaml#L1-L46) to comments.

Reply:
Done.

Reply:
The initial conclusion of this thread was to move Namespace into `InternalCertManagement` as `WebhookServiceNamespace`

Reply:
@alculquicondor 
As described in [this](https://github.com/kubernetes-sigs/kueue/pull/362#discussion_r961557630) comment by @kerthcet, I and @kerthcet  updated the conclusion that we should include the namespace in struct `Configuration` because it represents the namespace where kueue is deployed.

> Actually the namespace helps to control where to deploy the kueue components, besides configure this field, we should also update the config/default/kustomization.yaml, to change the value of namespace.
> 
> So I think we can leave it outside the internalCertManagement